pkg/ipam-controller/migrator/migrator.go

```diff
@@ -0,0 +1,250 @@
+	logger := logr.FromContextOrDiscard(ctx).WithName("migrator")
+	if os.Getenv(EnvDisableMigration) != "" {
+		logger.Info(fmt.Sprintf("%s set, skip controller migration", EnvDisableMigration))
+		return nil
+	}
+
+	cmName := DefaultConfigMapName
+	if os.Getenv(EnvConfigMapName) != "" {
+		cmName = os.Getenv(EnvConfigMapName)
+	}
+	cmNamespace := poolNamespace
+	if os.Getenv(EnvConfigMapNamespace) != "" {
+		cmNamespace = os.Getenv(EnvConfigMapNamespace)
+	}
+
+	cfg := &corev1.ConfigMap{}
+	key := types.NamespacedName{
+		Name:      cmName,
+		Namespace: cmNamespace,
+	}
+	err := c.Get(ctx, key, cfg)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			logger.Info("ConfigMap not found, skipping migration")
+			return nil
+		}
+		logger.Error(err, "failed to read ConfigMap object")
+		return err
+	}
+
+	confData, exist := cfg.Data[config.ConfigMapKey]
+	if !exist {
+		err = fmt.Errorf("invalid configuration: ConfigMap %s doesn't contain %s key",
+			key, config.ConfigMapKey)
+		logger.Error(err, "Invalid config, no data")
+		return err
+	}
+	controllerConfig := &config.Config{}
+	if err := json.Unmarshal([]byte(confData), controllerConfig); err != nil {
+		logger.Error(err, fmt.Sprintf("invalid configuration: ConfigMap %s contains invalid JSON",
+			config.ConfigMapKey))
+		return err
+	}
+	if err := controllerConfig.Validate(); err != nil {
+		logger.Error(err, fmt.Sprintf("invalid configuration: ConfigMap %s contains invalid config",
+			config.ConfigMapKey))
+		return err
+	}
+
+	pools := buildIPPools(controllerConfig, poolNamespace)
+
+	for name, p := range pools {
+		err = c.Create(ctx, p)
+		logger.Info(fmt.Sprintf("Creating IPPool: %v", p))
+		if apiErrors.IsAlreadyExists(err) {
+			existingPool := &ipamv1alpha1.IPPool{}
+			err = c.Get(ctx, client.ObjectKeyFromObject(p), existingPool)
+			if err != nil {
+				logger.Info("fail to get existing pool", "pool name", name)
+				return err
+			}
+			if !reflect.DeepEqual(existingPool.Spec, p.Spec) {
+				logger.Info("existing pool has different spec than config map setting", "pool name", name)
+				return fmt.Errorf("existing pool has different spec than config map setting")
+			}
+		} else if err != nil {
+			logger.Info("fail to create pool", "pool name", name)
+			return err
+		}
+	}
+
+	err = updateAllocations(ctx, c, logger, pools, poolNamespace)
+	if err != nil {
+		return err
+	}
+
+	err = c.Delete(ctx, &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Namespace: cmNamespace, Name: cmName}})
+	if err != nil {
+		logger.Info("fail to delete nv-ipam config map")
+		return err
+	}
+
+	return nil
+}
+
+func updateAllocations(ctx context.Context, c client.Client,
+	logger logr.Logger, pools map[string]*ipamv1alpha1.IPPool, poolNamespace string) error {
+	nodeList := &corev1.NodeList{}
+	if err := c.List(ctx, nodeList); err != nil {
+		logger.Error(err, "failed to list nodes")
+		return err
+	}
+	nodesToClearAnnotation := sets.New[string]()
+	for poolName := range pools {
```

Comment:
should we skip pools that already have allocations ? i think we should.

from either previous migration attempt or other reasons.

Reply:
Done